File: app/src/main/java/com/drdisagree/colorblendr/config/RPrefs.java
package com.drdisagree.colorblendr.config;

import static android.content.Context.MODE_PRIVATE;
import static com.drdisagree.colorblendr.common.Const.EXCLUDED_PREFS_FROM_BACKUP;

import android.content.SharedPreferences;
import android.util.Log;

import androidx.annotation.NonNull;

import com.drdisagree.colorblendr.ColorBlendr;
import com.drdisagree.colorblendr.common.Const;

import java.io.IOException;
import java.io.InputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.OutputStream;
import java.util.HashMap;
import java.util.Map;

@SuppressWarnings("unused")
public class RPrefs {

    private static final String TAG = RPrefs.class.getSimpleName();

    public static SharedPreferences prefs = ColorBlendr.getAppContext().createDeviceProtectedStorageContext().getSharedPreferences(Const.SharedPrefs, MODE_PRIVATE);
    static SharedPreferences.Editor editor = prefs.edit();

    public static void putBoolean(String key, boolean val) {
        editor.putBoolean(key, val).apply();
    }

    public static void putInt(String key, int val) {
        editor.putInt(key, val).apply();
    }

    public static void putLong(String key, long val) {
        editor.putLong(key, val).apply();
    }

    public static void putFloat(String key, float val) {
        editor.putFloat(key, val).apply();
    }

    public static void putString(String key, String val) {
        editor.putString(key, val).apply();
    }

    public static boolean getBoolean(String key) {
        return prefs.getBoolean(key, false);
    }

    public static boolean getBoolean(String key, Boolean defValue) {
        return prefs.getBoolean(key, defValue);
    }

    public static int getInt(String key) {
        return prefs.getInt(key, 0);
    }

    public static int getInt(String key, int defValue) {
        return prefs.getInt(key, defValue);
    }

    public static long getLong(String key) {
        return prefs.getLong(key, 0);
    }

    public static long getLong(String key, long defValue) {
        return prefs.getLong(key, defValue);
    }

    public static float getFloat(String key) {
        return prefs.getFloat(key, 0);
    }

    public static float getFloat(String key, float defValue) {
        return prefs.getFloat(key, defValue);
    }

    public static String getString(String key) {
        return prefs.getString(key, null);
    }

    public static String getString(String key, String defValue) {
        return prefs.getString(key, defValue);
    }

    public static void clearPref(String key) {
        editor.remove(key).apply();
    }

    public static void clearPrefs(String... keys) {
        for (String key : keys) {
            editor.remove(key).apply();
        }
    }

    public static void clearAllPrefs() {
        editor.clear().apply();
    }

    public static void backupPrefs(final @NonNull OutputStream outputStream) {
        try (outputStream; ObjectOutputStream objectOutputStream = new ObjectOutputStream(outputStream)) {
            objectOutputStream.writeObject(prefs.getAll());
        } catch (IOException e) {
            Log.e(TAG, "Error serializing preferences", e);
        }

        try (outputStream; ObjectOutputStream objectOutputStream = new ObjectOutputStream(outputStream)) {
            Map<String, ?> allPrefs = prefs.getAll();

            for (String excludedPref : EXCLUDED_PREFS_FROM_BACKUP) {
                allPrefs.remove(excludedPref);
            }

            objectOutputStream.writeObject(allPrefs);
        } catch (IOException e) {
            Log.e(TAG, "Error serializing preferences", e);
        }
    }

    @SuppressWarnings("unchecked")
    public static void restorePrefs(final @NonNull InputStream inputStream) {
        Map<String, Object> map;
        try (inputStream; ObjectInputStream objectInputStream = new ObjectInputStream(inputStream)) {
            map = (Map<String, Object>) objectInputStream.readObject();
        } catch (ClassNotFoundException | IOException e) {
            Log.e(TAG, "Error deserializing preferences", e);
            return;
        }

        // Retrieve excluded prefs from current prefs
        Map<String, Object> excludedPrefs = new HashMap<>();
        for (String excludedPref : EXCLUDED_PREFS_FROM_BACKUP) {
            Object prefValue = prefs.getAll().get(excludedPref);
            if (prefValue != null) {
                excludedPrefs.put(excludedPref, prefValue);
            }
        }

        editor.clear();

        // Restore excluded prefs
        for (Map.Entry<String, Object> entry : excludedPrefs.entrySet()) {
            putObject(entry.getKey(), entry.getValue());
        }

        // Restore non-excluded prefs
        for (Map.Entry<String, Object> entry : map.entrySet()) {
            if (EXCLUDED_PREFS_FROM_BACKUP.contains(entry.getKey())) {
                continue;
            }

            putObject(entry.getKey(), entry.getValue());
        }

        editor.apply();
    }

    private static void putObject(String key, Object value) {
        if (value instanceof Boolean) {
            editor.putBoolean(key, (Boolean) value);
        } else if (value instanceof String) {
            editor.putString(key, (String) value);
        } else if (value instanceof Integer) {
            editor.putInt(key, (Integer) value);
        } else if (value instanceof Float) {
            editor.putFloat(key, (Float) value);
        } else if (value instanceof Long) {
            editor.putLong(key, (Long) value);
        } else {
            throw new IllegalArgumentException("Type " + value.getClass().getName() + " is unknown");
        }
    }
}

File: app/src/main/java/com/drdisagree/colorblendr/config/RPrefs.kt
package com.drdisagree.colorblendr.config

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import com.drdisagree.colorblendr.ColorBlendr.Companion.appContext
import com.drdisagree.colorblendr.common.Const
import com.drdisagree.colorblendr.common.Const.EXCLUDED_PREFS_FROM_BACKUP
import com.drdisagree.colorblendr.common.Const.MONET_SEED_COLOR
import com.drdisagree.colorblendr.common.Const.MONET_SEED_COLOR_ENABLED
import com.drdisagree.colorblendr.common.Const.SAVED_CUSTOM_MONET_STYLES
import com.drdisagree.colorblendr.common.Const.THEMING_ENABLED
import com.drdisagree.colorblendr.common.Const.WALLPAPER_COLOR_LIST
import com.google.gson.reflect.TypeToken
import java.io.IOException
import java.io.InputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.io.OutputStream

@Suppress("unused")
object RPrefs {
    private val TAG: String = RPrefs::class.java.simpleName

    private var prefs: SharedPreferences = appContext
        .createDeviceProtectedStorageContext()
        .getSharedPreferences(Const.SHARED_PREFS, Context.MODE_PRIVATE)
    private var editor: SharedPreferences.Editor = prefs.edit()

    fun putBoolean(key: String, `val`: Boolean) = editor.putBoolean(key, `val`).apply()

    fun putInt(key: String, `val`: Int) = editor.putInt(key, `val`).apply()

    fun putLong(key: String, `val`: Long) = editor.putLong(key, `val`).apply()

    fun putFloat(key: String, `val`: Float) = editor.putFloat(key, `val`).apply()

    fun putString(key: String, `val`: String) = editor.putString(key, `val`).apply()

    fun getBoolean(key: String): Boolean = prefs.getBoolean(key, false)

    fun getBoolean(key: String, defValue: Boolean): Boolean = prefs.getBoolean(key, defValue)

    fun getInt(key: String): Int = prefs.getInt(key, 0)

    fun getInt(key: String, defValue: Int): Int = prefs.getInt(key, defValue)

    fun getLong(key: String): Long = prefs.getLong(key, 0)

    fun getLong(key: String, defValue: Long): Long = prefs.getLong(key, defValue)

    fun getFloat(key: String): Float = prefs.getFloat(key, 0f)

    fun getFloat(key: String, defValue: Float): Float = prefs.getFloat(key, defValue)

    fun getString(key: String): String? = prefs.getString(key, null)

    fun getString(key: String, defValue: String?): String? = prefs.getString(key, defValue)

    fun getAllPrefs(): MutableMap<String, *> = prefs.all

    fun getAllPrefsAsGson(): String = Const.GSON.toJson(getAllPrefs() as Map<String, *>)

    fun Map<String, Any>.toGsonString(): String = Const.GSON.toJson(this)

    fun String.toPrefs(): Map<String, Any> =
        Const.GSON.fromJson(this, object : TypeToken<Map<String, Any>>() {}.type)

    fun clearPref(key: String) = editor.remove(key).apply()

    fun clearPrefs(vararg keys: String) = keys.forEach { key -> editor.remove(key).apply() }

    fun clearAllPrefs() = editor.clear().apply()

    fun backupPrefs(outputStream: OutputStream) {
        try {
            outputStream.use {
                ObjectOutputStream(outputStream).use { objectOutputStream ->
                    val allPrefs = getAllPrefs()
                    EXCLUDED_PREFS_FROM_BACKUP.forEach { excludedPref ->
                        allPrefs.remove(excludedPref)
                    }
                    objectOutputStream.writeObject(allPrefs)
                }
            }
        } catch (e: IOException) {
            Log.e(TAG, "Error serializing preferences", e)
        }
    }

    @Suppress("UNCHECKED_CAST")
    fun restorePrefs(inputStream: InputStream) {
        var map: Map<String, Any>
        try {
            inputStream.use {
                ObjectInputStream(inputStream).use { objectInputStream ->
                    map = objectInputStream.readObject() as Map<String, Any>
                }
            }
        } catch (e: ClassNotFoundException) {
            Log.e(TAG, "Error deserializing preferences", e)
            return
        } catch (e: IOException) {
            Log.e(TAG, "Error deserializing preferences", e)
            return
        }

        restorePrefsMap(map)
    }

    fun restorePrefsMap(map: Map<String, Any>, isApplyingTheme: Boolean = false) {
        val allPrefs = getAllPrefs()

        // Retrieve excluded prefs from current prefs
        val excludedPrefs: MutableMap<String, Any> = HashMap()

        // Restoring config will enable theming service
        excludedPrefs[THEMING_ENABLED] = true

        EXCLUDED_PREFS_FROM_BACKUP.forEach { excludedPref ->
            val prefValue = allPrefs[excludedPref]
            if (prefValue != null) {
                excludedPrefs[excludedPref] = prefValue
            }
        }

        // Restore saved themes when enabling a theme otherwise it will be
        // overwritten by the new preference of the theme
        val savedThemes = allPrefs[SAVED_CUSTOM_MONET_STYLES]
        if (savedThemes != null && isApplyingTheme) {
            excludedPrefs[SAVED_CUSTOM_MONET_STYLES] = allPrefs[SAVED_CUSTOM_MONET_STYLES] as String
        }

        // Check if seed color is available in current wallpaper color list
        val seedColor = map[MONET_SEED_COLOR] as? Int
        val wallpaperColors = allPrefs[WALLPAPER_COLOR_LIST] as? String
        val colorAvailable = if (seedColor != null && wallpaperColors != null) {
            Const.GSON.fromJson<ArrayList<Int?>?>(
                wallpaperColors,
                object : TypeToken<ArrayList<Int?>?>() {}.type
            )?.contains(seedColor) ?: false
        } else false

        editor.clear()

        // Restore excluded prefs
        for ((key, value) in excludedPrefs) {
            putObject(key, value)
        }

        // Restore non-excluded prefs
        for ((key, value) in map) {
            if (EXCLUDED_PREFS_FROM_BACKUP.contains(key) ||
                (isApplyingTheme && key == SAVED_CUSTOM_MONET_STYLES)
            ) {
                continue
            }

            putObject(key, value)
        }

        // Set basic color if seed color is not listed in wallpaper colors
        putObject(MONET_SEED_COLOR_ENABLED, !colorAvailable)

        editor.apply()
    }

    private fun putObject(key: String, value: Any) {
        when (value) {
            is Boolean -> editor.putBoolean(key, value)
            is String -> editor.putString(key, value)
            is Int -> editor.putInt(key, value)
            is Long -> editor.putLong(key, value)
            // Float and Double are unused in this project
            // is Float -> editor.putFloat(key, value)
            // is Double -> editor.putFloat(key, value.toFloat())
            is Float -> editor.putInt(key, value.toInt())
            is Double -> editor.putInt(key, value.toInt())
            else -> throw IllegalArgumentException("Type ${value.javaClass.simpleName} is unknown")
        }
    }
}